$NetBSD: patch-ad,v 1.9 2008/07/03 21:06:10 tonnerre Exp $

Avoid memory size integer overflow memory exhaustion DoS in filling
arrays (SN-2008-02).

--- array.c.orig	2008-07-03 22:56:32.000000000 +0200
+++ array.c
@@ -2416,10 +2416,10 @@ rb_ary_fill(argc, argv, ary)
 	break;
     }
     rb_ary_modify(ary);
-    end = beg + len;
-    if (end < 0) {
+    if (beg >= ARY_MAX_SIZE || len > ARY_MAX_SIZE - beg) {
 	rb_raise(rb_eArgError, "argument too big");
     }
+    end = beg + len;
     if (end > RARRAY(ary)->len) {
 	if (end >= RARRAY(ary)->aux.capa) {
 	    REALLOC_N(RARRAY(ary)->ptr, VALUE, end);
